coffea/analysis_tools.py

```diff
@@ -118,6 +118,82 @@ def add(self, name, weight, weightUp=None, weightDown=None, shift=False):
+            modifierNames: list of str
+                list of modifiers for each set of weights variation
+            weightsUp : list of numpy.ndarray
+                weight with correction uncertainty shifted up (if available)
+            weightsDown : list of numpy.ndarray
+                weight with correction uncertainty shifted down. If ``weightUp`` is supplied, and
+                the correction uncertainty is symmetric, this can be set to None to auto-calculate
+                the down shift as ``1 / weightUp``.
+            shift : bool, optional
+                if True, interpret weightUp and weightDown as a realtive difference (additive) to the
+                nominal value
+
+        .. note:: ``weightUp`` and ``weightDown`` are assumed to be rvalue-like and may be modified in-place by this function
+        """
+        if name.endswith("Up") or name.endswith("Down"):
+            raise ValueError(
+                "Avoid using 'Up' and 'Down' in weight names, instead pass appropriate shifts to add() call"
+            )
+        weight = coffea.util._ensure_flat(weight, allow_missing=True)
+        if isinstance(weight, numpy.ma.MaskedArray):
+            # TODO what to do with option-type? is it representative of unknown weight
+            # and we default to one or is it an invalid weight and we should never use this
+            # event in the first place (0) ?
+            weight = weight.filled(1.0)
+        self._weight = self._weight * weight
+        if self._storeIndividual:
+            self._weights[name] = weight
+        # Now loop on the variations
+        if len(modifierNames) > 0:
+            if len(modifierNames) != len(weightsUp) or len(modifierNames) != len(
+                weightsDown
+            ):
+                raise ValueError(
+                    "Provide the same number of modifier names related to the list of modified weights"
+                )
+        for modifier, weightUp, weightDown in zip(
+            modifierNames, weightsUp, weightsDown
+        ):
+            if weightUp is not None:
+                weightUp = coffea.util._ensure_flat(weightUp, allow_missing=True)
+                if isinstance(weightUp, numpy.ma.MaskedArray):
+                    weightUp = weightUp.filled(1.0)
+                if shift:
+                    weightUp += weight
+                weightUp[weight != 0.0] /= weight[weight != 0.0]
+                self._modifiers[name + "_" + modifier + "Up"] = weightUp
+            if weightDown is not None:
+                weightDown = coffea.util._ensure_flat(weightDown, allow_missing=True)
+                if isinstance(weightDown, numpy.ma.MaskedArray):
+                    weightDown = weightDown.filled(1.0)
+                if shift:
+                    weightDown = weight - weightDown
+                weightDown[weight != 0.0] /= weight[weight != 0.0]
+                self._modifiers[name + "_" + modifier + "Down"] = weightDown
```

Comment:
It would be nice to break this almost-duplication of L97-112 into a helper function.